packages/core/src/api/positionMapping.ts

```diff
@@ -0,0 +1,145 @@
+ */
+export class PositionStorage<
+  BSchema extends BlockSchema = DefaultBlockSchema,
+  ISchema extends InlineContentSchema = DefaultInlineContentSchema,
+  SSchema extends StyleSchema = DefaultStyleSchema
+> {
+  private readonly editor: BlockNoteEditor<BSchema, ISchema, SSchema>;
+  /**
+   * Whether the editor has had a remote transaction.
+   */
+  private hadRemoteTransaction = false;
+
+  private readonly mapping = new Mapping();
+  private mappingLength = 0;
+
+  constructor(
+    editor: BlockNoteEditor<BSchema, ISchema, SSchema>,
+    { shouldMount = true }: { shouldMount?: boolean } = {}
+  ) {
+    this.editor = editor;
+    this.onTransactionHandler = this.onTransactionHandler.bind(this);
+
+    if (!shouldMount) {
+      return;
+    }
+
+    if (!this.editor._tiptapEditor) {
+      throw new Error("Editor not mounted");
+    }
+    this.editor._tiptapEditor.on("transaction", this.onTransactionHandler);
+  }
+
+  /**
+   * This will be called whenever a transaction is applied to the editor.
+   *
+   * It's used to update the position mapping or tell if there was a remote transaction.
+   */
+  private onTransactionHandler({ transaction }: { transaction: Transaction }) {
+    if (this.hadRemoteTransaction) {
+      // If we have already had a remote transaction, we rely only on relative positions, so no need to update the mapping.
+      return;
+    }
+
+    if (isRemoteTransaction(transaction)) {
+      this.hadRemoteTransaction = true;
+    } else {
+      this.mapping.appendMapping(transaction.mapping);
+      this.mappingLength += transaction.mapping.maps.length;
+    }
+  }
+
+  /**
+   * This is used to track a position in the editor.
+   *
+   * @param position The position to track.
+   * @param side The side of the position to track. "left" is the default. "right" would move with the change if the change is in the right direction.
+   * @param getOffset This allows you to offset the returned position from the tracked position. This is useful for cases where the tracked position is not the actual position of the element.
+   */
+  public track(
+    /**
+     * The position to track.
+     */
+    position: number,
+    /**
+     * This is the side of the position to track. "left" is the default. "right" would move with the change if the change is in the right direction.
+     */
+    side: "left" | "right" = "left"
+  ): () => number | undefined {
+    const ySyncPluginState = ySyncPluginKey.getState(
+      this.editor._tiptapEditor.state
+    ) as YSyncPluginState;
+
+    const trackedMapLength = this.mappingLength;
+    if (!ySyncPluginState) {
+      return () => {
+        const pos = this.mapping
+          .slice(trackedMapLength)
+          .map(position, side === "left" ? -1 : 1);
+
+        return pos;
+      };
+    }
+
+    const relativePosition = absolutePositionToRelativePosition(
```

Comment:
apparently yjs also supports `assoc` for left / right: https://github.com/yjs/yjs/blob/0e3195797fd6fce39b987f7aebd2529681f29e15/src/utils/RelativePosition.js#L163 - but it's not available in the prosemiror "absToRel" functions.

Could this cause any bugs (the fact that we now just do `rel-1` at the end?)

Reply:
The yjs implementation is just subtracting on the index (because yjs normally does right association since that is how most writing is done & compresses well for runs). So, this should be exactly equivalent.